File: tests/core/matrices/codac2_tests_IntFullPivLU.cpp
/** 
 *  Codac tests
 * ----------------------------------------------------------------------------
 *  \date       2024
 *  \author     Damien Massé
 *  \copyright  Copyright 2024 Codac Team
 *  \license    GNU Lesser General Public License (LGPL)
 */

#include <catch2/catch_test_macros.hpp>
#include <codac2_IntvFullPivLU.h>
#include <codac2_BoolInterval.h>
#include <codac2_IntervalVector.h>
#include <iostream>

using namespace std;
using namespace codac2;

TEST_CASE("IntvFullPivLU")
{
  /* ``classic'' invertible matrix */
  {
    Matrix M 
    { { 1, -4,  3, 7 },
      { 2,  1, -4, 6 },
      { 5,  2,  1, 9 },
      { -1, 0,  3, 2 } };
    IntvFullPivLU LUdec(M);
    CHECK(LUdec.is_injective()==BoolInterval::TRUE);
    CHECK(LUdec.is_surjective()==BoolInterval::TRUE);
    CHECK(LUdec.rank()==Interval(4));
    CHECK((LUdec.determinant()+602).mag()<=1e-10);
    CHECK((LUdec.reconstructed_matrix()-M).lpNorm<Eigen::Infinity>().ub()<=1e-10);
    IntervalMatrix I1 = LUdec.solve(IntervalMatrix::Identity(4,4));
    CHECK((IntervalMatrix(M)*I1-IntervalMatrix::Identity(4,4))
  		.lpNorm<Eigen::Infinity>().ub()<=1e-10);
  }
  /* singular matrix , but approximations may make it non singular */
  {
    Matrix M 
    { { 1, -4, 6, 7 },
      { 2,  1, 3, 6 },
      { 5,  2, 8, 9 },
      { -1, 0, -2, 2 } };
    IntvFullPivLU LUdec(M);
    CHECK(LUdec.is_injective()==BoolInterval::UNKNOWN);
    CHECK(LUdec.is_surjective()==BoolInterval::UNKNOWN);
    CHECK(LUdec.rank()==Interval(3,4));
    CHECK(LUdec.determinant().mag()<=1e-10);
    CHECK((LUdec.reconstructed_matrix()-M).lpNorm<Eigen::Infinity>().ub()<=1e-10);
    IntervalMatrix K = LUdec.kernel();
    CHECK(K.cols()==1);
    CHECK((IntervalMatrix(M)*K).lpNorm<Eigen::Infinity>().ub()<=1e-10);
    IntervalMatrix coK = LUdec.cokernel();
    CHECK(coK.rows()==1);
    CHECK((coK*IntervalMatrix(M)).lpNorm<Eigen::Infinity>().ub()<=1e-10);
  }
  
  /* rectangular matrix, full rank */
  {
    Matrix M 
    { { 1, -4, 6, 7, 6 },
      { 2,  1, 3, 6, -2 },
      { 5,  2, 2, 9, -1 } };
    IntvFullPivLU LUdec(M);
    CHECK(LUdec.is_injective()==BoolInterval::FALSE);
    CHECK(LUdec.is_surjective()==BoolInterval::TRUE);
    CHECK(LUdec.rank()==Interval(3));
    CHECK((LUdec.reconstructed_matrix()-M).lpNorm<Eigen::Infinity>().ub()<=1e-10);
    IntervalMatrix K = LUdec.kernel();
    CHECK(K.cols()==2);
    CHECK((IntervalMatrix(M)*K).lpNorm<Eigen::Infinity>().ub()<=1e-10);
    IntervalMatrix Img = LUdec.image(M);
    CHECK(Img.cols()==3);
    IntervalMatrix I1 = LUdec.solve(IntervalMatrix::Identity(3,3));
    CHECK((IntervalMatrix(M)*I1-IntervalMatrix::Identity(3,3))
  		.lpNorm<Eigen::Infinity>().ub()<=1e-10);
    IntervalMatrix A { { 1 }, 
		       { {-20,20} },
		       { 2 },
		       { {-20,20} },
		       { {-20,20} }};
    IntervalVector B { 2.0,1.0,4.0 };
    LUdec.solve(B,A);
    CHECK((IntervalMatrix(M)*A-B)
  		.lpNorm<Eigen::Infinity>().ub()<=1e-10);
  }
  
  /* rectangular matrix, full rank */
  {
    Matrix M 
    { { 1, -4, 6 },
      { 2,  1, 3 },
      { 5,  2, 2 },
      { 1,  3, -3},
      { 0,  0, 1 } };
    IntvFullPivLU LUdec(M);
    CHECK(LUdec.is_injective()==BoolInterval::TRUE);
    CHECK(LUdec.is_surjective()==BoolInterval::FALSE);
    CHECK(LUdec.rank()==Interval(3));
    IntervalMatrix CoImg = LUdec.coimage(M);
    CHECK(CoImg.rows()==3);
    IntervalMatrix coK = LUdec.cokernel();
    CHECK(coK.rows()==2);
    CHECK((coK*IntervalMatrix(M)).lpNorm<Eigen::Infinity>().ub()<=1e-10);
    CHECK((LUdec.reconstructed_matrix()-M).lpNorm<Eigen::Infinity>().ub()<=1e-10);
  }
  
  
  /* interval matrix, full rank */
  {
    IntervalMatrix M 
    { { {-4,0}, 1, {5,6}, 7, 6 },
      { {1,3}, 2, 3, {3,6}, -2 },
      { {0,2}, 5, 2, 9, {-4,-1} } };
    IntvFullPivLU LUdec(M);
    CHECK(M.is_subset(LUdec.reconstructed_matrix()));
    CHECK(LUdec.is_injective()==BoolInterval::FALSE);
    CHECK(LUdec.is_surjective()==BoolInterval::TRUE);
    CHECK(LUdec.rank()==Interval(3));
    IntervalMatrix K = LUdec.kernel();
    CHECK(K.cols()==2);
    IntervalMatrix Img = LUdec.image(M);
    CHECK(Img.cols()==3);
  }
  
  /* interval matrix,  not full rank */
  {
    IntervalMatrix M 
    { { {0.1,0.2}, {0.1,0.2}, {0.1,0.2}, {0.1,0.2}, {0.1,0.2} },
      { {0.1,0.2}, {0.1,0.2}, {0.1,0.2}, {0.1,0.2}, {0.1,0.2} },
      { {0.1,0.2}, {0.1,0.2}, {0.1,0.2}, {0.1,0.2}, {0.1,0.2} } };
    IntvFullPivLU LUdec(M);
    CHECK(M.is_subset(LUdec.reconstructed_matrix()));
    CHECK(LUdec.is_injective()==BoolInterval::FALSE);
    CHECK(LUdec.is_surjective()==BoolInterval::UNKNOWN);
    CHECK(LUdec.rank()==Interval(1,3));
    IntervalMatrix K = LUdec.kernel();
    CHECK(K.cols()==4);
    IntervalMatrix K2 = LUdec.cokernel();
    CHECK(K2.rows()==2);
    IntervalMatrix Img = LUdec.image(M);
    CHECK(Img.cols()==1);
    IntervalMatrix CoImg = LUdec.coimage(M);
    CHECK(CoImg.rows()==1);
  }
  
}

TEST_CASE("IntvFullPivLU solve tall matrix")
{
  Matrix A({
    { 1, 0 },
    { 0, 1 },
    { 1, 1 },
    { 2,-1 },
    {-1, 2 },
  });

  Matrix X({
    { 2,-1 },
    {-1, 3 },
  });

  Matrix rhs_mid = A*X;
  IntervalMatrix rhs(rhs_mid);

  IntvFullPivLU lu(A);

  SECTION("solve(rhs) with rows >= cols + 2")
  {
    IntervalMatrix sol = lu.solve(rhs);

    CHECK(!sol.is_empty());
    CHECK(sol.contains(X));
    CHECK((A.template cast<Interval>() * sol).contains(rhs_mid));
  }

  SECTION("solve(rhs, B) with rows >= cols + 2")
  {
    IntervalMatrix B = IntervalMatrix::Constant(2,2,{-10.0,10.0});

    lu.solve(rhs, B);

    CHECK(!B.is_empty());
    CHECK(B.contains(X));
    CHECK((A.template cast<Interval>() * B).contains(rhs_mid));
  }

  SECTION("solve(rhs) detects inconsistent rhs")
  {
    Matrix rhs_bad_mid = rhs_mid;
    rhs_bad_mid(4,0) += 1.0; // breaks consistency on the extra row

    IntervalMatrix rhs_bad(rhs_bad_mid);
    IntervalMatrix sol = lu.solve(rhs_bad);

    CHECK(sol.is_empty());
  }

  SECTION("solve(rhs, B) detects inconsistent rhs")
  {
    Matrix rhs_bad_mid = rhs_mid;
    rhs_bad_mid(4,0) += 1.0; // breaks consistency on the extra row

    IntervalMatrix rhs_bad(rhs_bad_mid);
    IntervalMatrix B = IntervalMatrix::Constant(2,2,{-10.0,10.0});

    lu.solve(rhs_bad, B);

    CHECK(B.is_empty());
  }
}

File: tests/core/matrices/codac2_tests_IntFullPivLU.py
#!/usr/bin/env python

#  Codac tests
# ----------------------------------------------------------------------------
#  \date       2024
#  \author     Damien Massé
#  \copyright  Copyright 2024 Codac Team
#  \license    GNU Lesser General Public License (LGPL)


import unittest
from codac import *

class TestIntvFullPivLU(unittest.TestCase):

  def test_IntvFullPivLU_1(self):
    M = Matrix([
    [ 1, -4, 3, 7 ],
    [ 2, 1, -4, 6 ],
    [ 5, 2, 1 , 9 ],
    [ -1, 0, 3, 2 ]
    ])

    LUdec = IntvFullPivLU(M)
    self.assertTrue(LUdec.is_injective()==BoolInterval.TRUE)
    self.assertTrue(LUdec.is_surjective()==BoolInterval.TRUE)
    self.assertTrue(LUdec.rank()==Interval(4))
    self.assertTrue((LUdec.determinant()+602).mag()<=1e-10)
    self.assertTrue((LUdec.reconstructed_matrix()-M).norm().ub()<=1e-10)
    I1 = LUdec.solve(IntervalMatrix.eye(4,4));
    self.assertTrue((I1*M-Matrix.eye(4,4)).norm().ub()<1e-10)

  def test_IntvFullPivLU_2(self):
    M = Matrix([
    [ 1, -4, 6, 7 ],
    [ 2, 1, 3, 6 ],
    [ 5, 2, 8 , 9 ],
    [ -1, 0, -2, 2 ]
    ])

    LUdec = IntvFullPivLU(M)
    self.assertTrue(LUdec.is_injective()==BoolInterval.UNKNOWN)
    self.assertTrue(LUdec.is_surjective()==BoolInterval.UNKNOWN)
    self.assertTrue(LUdec.rank()==Interval([3,4]))
    self.assertTrue((LUdec.determinant()).mag()<=1e-10)
    self.assertTrue((LUdec.reconstructed_matrix()-M).norm().ub()<=1e-10)
    K = LUdec.kernel()
    self.assertTrue(K.cols()==1)
    self.assertTrue((M*K).norm().ub()<1e-10)
    coK = LUdec.cokernel()
    self.assertTrue(coK.rows()==1)
    self.assertTrue((coK*M).norm().ub()<1e-10)
    Im = LUdec.image(M)
    self.assertTrue(Im.cols()==3)
    coIm = LUdec.coimage(M)
    self.assertTrue(coIm.rows()==3)

  def test_IntvFullPivLU_3(self):
    M = Matrix([
    [ 1, -4, 6, 7, 6 ],
    [ 2, 1, 3, 6, -2 ],
    [ 5, 2, 8 , 9, -1 ]
    ])
    LUdec = IntvFullPivLU(M)
    self.assertTrue(LUdec.is_injective()==BoolInterval.FALSE)
    self.assertTrue(LUdec.is_surjective()==BoolInterval.TRUE)
    self.assertTrue(LUdec.rank()==Interval(3))
    self.assertTrue((LUdec.reconstructed_matrix()-M).norm().ub()<=1e-10)
    K = LUdec.kernel()
    self.assertTrue(K.cols()==2)
    self.assertTrue((M*K).norm().ub()<1e-10)
    Im = LUdec.image(M)
    self.assertTrue(Im.cols()==3)
    I1 = LUdec.solve(IntervalMatrix.eye(3,3));
    self.assertTrue((M*I1-Matrix.eye(3,3)).norm().ub()<1e-10)
    A = IntervalMatrix([ [1], [[-20,20]], [2], [[-20,20]], [[-20,20]] ])
    B = IntervalMatrix([ [2.0], [1.0], [4.0] ])
    LUdec.solve(B,A)
    self.assertTrue((M*A-B).norm().ub()<=1e-10)

  def test_IntvFullPivLU_solve_tall_matrix(self):

    A = Matrix([
     [ 1, 0 ],
     [ 0, 1 ],
     [ 1, 1 ],
     [ 2,-1 ],
     [-1, 2 ],
    ])
 
    X = Matrix([
     [ 2,-1 ],
     [-1, 3 ],
    ])
 
    rhs_mid = A*X
    rhs = IntervalMatrix(rhs_mid)
 
    lu = IntvFullPivLU(A)
 
    sol = lu.solve(rhs)
 
    self.assertTrue(not sol.is_empty())
    self.assertTrue(sol.contains(X))
    self.assertTrue((A*sol).contains(rhs_mid))
 
  def test_IntvFullPivLU_solve_tall_matrix_with_box(self):

    A = Matrix([
     [ 1, 0 ],
     [ 0, 1 ],
     [ 1, 1 ],
     [ 2,-1 ],
     [-1, 2 ],
    ])
 
    X = Matrix([
     [ 2,-1 ],
     [-1, 3 ],
    ])
 
    rhs_mid = A*X
    rhs = IntervalMatrix(rhs_mid)
 
    lu = IntvFullPivLU(A)
    B = IntervalMatrix.constant(2,2,[-10,10])
 
    lu.solve(rhs, B)
 
    self.assertTrue(not B.is_empty())
    self.assertTrue(B.contains(X))
    self.assertTrue((A*B).contains(rhs_mid))

  def test_IntvFullPivLU_solve_tall_matrix_inconsistent_rhs(self):

    A = Matrix([
     [ 1, 0 ],
     [ 0, 1 ],
     [ 1, 1 ],
     [ 2,-1 ],
     [-1, 2 ],
    ])
 
    rhs_bad = IntervalMatrix([
      [[ 2, 2],[-1,-1]],
      [[-1,-1],[ 3, 3]],
      [[ 1, 1],[ 2, 2]],
      [[ 5, 5],[-5,-5]],
      [[-3,-3],[ 7, 7]], # should be [-4, 7] for consistency
    ])
 
    lu = IntvFullPivLU(A)
 
    sol = lu.solve(rhs_bad)
    self.assertTrue(sol.is_empty())
 
    B = IntervalMatrix.constant(2,2,[-10,10])
    lu.solve(rhs_bad, B)
    self.assertTrue(B.is_empty())
    
if __name__ ==  '__main__':
  unittest.main()